datafusion/common/src/tree_node.rs

```diff
@@ -0,0 +1,261 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! This module provides common traits for visiting or rewriting tree nodes easily.
+
+use crate::{DataFusionError, Result};
+
+/// Trait for tree node. It can be [`ExecutionPlan`], [`PhysicalExpr`], [`LogicalExpr`], etc.
+pub trait TreeNode: Clone {
+    /// Return the children of this tree node
+    fn get_children(&self) -> Vec<Self>;
+
+    /// Use preorder to iterate the node on the tree so that we can stop fast for some cases.
+    ///
+    /// `op` can be used to collect some info from the tree node.
+    fn collect<F>(&self, op: &mut F) -> Result<()>
```

Comment:
`collect` usually refers to apply some func to each element and the returned result should be a container type.

Reply:
Thanks for your suggestion. Maybe it's better to change the name to `apply`.

Reply:
May be `try_for_each` ?

Reply:
Personally, I prefer to use `apply`. And later I'll remove the `get_children` and add `apply_children`.